tests/system/source/main.d

```diff
@@ -38,127 +38,93 @@ import core.time;
-            assert(height == 0);
-        }
-
-        // Make sure the nodes use the test genesis block
-        const blocks = clients[0].getBlocksFrom(0, 1);
-        assert(blocks.length == 1);
-        assert(blocks[0] == TestGenesis.GenesisBlock);
-
-        auto kp = WK.Keys.Genesis;
-
-        writefln("%s Put 1 transaction to client[0]", PREFIX);
-        genesisSpendable().takeExactly(1)
-            .map!(txb => txb.sign()).each!(tx => clients[0].putTransaction(tx));
-
-        writefln("%s Wait 5 seconds", PREFIX);
-        Thread.sleep(5.seconds); // Give time for block to be externalized
-
-        writefln("%s Check height is 1", PREFIX);
-        assertBlockHeight(clients, addresses, 1);
-
-        // wait for distance 1 to be revealed before block 2 is proposed
-        waitForPreimages(clients, addresses, 1, 5.seconds);
-        writefln("%s Put 8 transactions to client[0]", PREFIX);
-        genesisSpendable().dropExactly(1).takeExactly(1)
-            .map!(txb =>
-                txb.split(WK.Keys.byRange.map!(k => k.address)
-                    .take(8)).sign())
-            .each!(tx => clients[0].putTransaction(tx));
-
-        writefln("%s Wait 5 seconds", PREFIX);
-        Thread.sleep(5.seconds); // Give time for block to be externalized
+    API[] clients;
+    foreach (const ref addr; addresses)
+        clients ~= new RestInterfaceClient!API(addr);
 
-        writefln("%s Check height is 2", PREFIX);
-        assertBlockHeight(clients, addresses, 2);
+    // Get node name from address for logging
+    string nodeFromClientIndex(size_t index)
+    {
+        return format!"node-%s"(addresses[index][$-1 .. $]);
     }
 
-    return 0;
-}
-
-/// wait for preimages for the given distance to be revealed
-private void waitForPreimages (API[] clients, in string[] addresses, in uint distance,
-    Duration timeout)
-{
-    clients.each!(client =>
-        TestGenesis.GenesisBlock.header.enrollments.each!(enroll =>
-            retryFor(client.getPreimage(enroll.utxo_key)
-                .distance >= distance, timeout)));
-}
+    // keep polling the nodes for a complete network discovery, until a timeout
+    writefln("%s waitForDiscovery", PREFIX);
+    const discovery_duration = 20.seconds;
+    clients.enumerate.each!((idx, client) =>
+    {
+        retryFor(client.getNodeInfo().ifThrown(NodeInfo.init)
+            .state == NetworkState.Complete,
+            discovery_duration,
+            format("%s %s has not completed discovery after %s.",
+                PREFIX, nodeFromClientIndex(idx), discovery_duration * (idx + 1)));
+    }());
 
-/// Check block generation
-private void assertBlockHeight (API[] clients, const string[] addresses, ulong height)
-{
-    Hash blockHash;
-    size_t times; // Number of times we slept for 500 msecs
-    foreach (idx, ref client; clients)
+    /// Check block generation
+    void assertBlockHeightAtleast (ulong height)
```

Comment:
Don't think we need  the `AtLeast` either. We just check that a certain block height has been externalized, not that it's the latest.

Reply:
In some tests (in fact the previous implementation) we expected the exact height as we had control of the txs sent. I think it is better to indicate that we don't care if more blocks have been created than our target.